### PR TITLE
Contributions Tab (parity with profile)

### PR DESCRIPTION
**Pull request by twrichards**
Replicates and improves upon the contributions tab from profile.
image

## Key Feature : Amount Updater
NOW uses the bounds from https://github.com/guardian/support-frontend/blob/5aa237d/assets/helpers/contributions.js which differ based on currency and  frequency (monthly vs annual) and also has an improved 'sense check' where you are asked to double check if you increase your contribution by x10 (indicative of an accidental extra key press).

image

image

image

image

image

image

image

image

...or, if it fails...

image

### CODE REVIEW

**Comment by twrichards** [comment]
Original amount updater implementation on profile : https://github.com/guardian/frontend/pull/18506

**Comment by twrichards** [comment]
Tested in IE11 and on iPhone6S, via BrowserStack

File: app/client/components/payment/update/stripeCardInputForm.tsx [hunk 1]
```diff
@@ -21,7 +20,6 @@ export interface StripeCardInputFormProps
 
 export interface StripeCardInputFormState {
   isGeneratingToken: boolean;
-  isValid: boolean;
```
Comment (twrichards) [inline app/client/components/payment/update/stripeCardInputForm.tsx]: this was an un-used 'state' variable so removed

File: app/server/server.ts [hunk 2]
```diff
@@ -156,6 +160,22 @@ Object.values(ProductTypes).forEach((productType: ProductType) => {
         })()
     )
   );
+
+  server.use(
```
Comment (twrichards) [inline app/server/server.ts]: moved banner redirect into the dynamic routes (one per ProductType) rather than being a broad on any string

**Comment by ionamckendrick** [comment]
A small thing, but could we switch to "a month" rather than "per month". Not a blocker, but if you were happy to change ... Otherwise it's looking good!

**Comment by twrichards** [comment]
> A small thing, but could we switch to "a month" rather than "per month". Not a blocker, but if you were happy to change ... Otherwise it's looking good!

@ionamckendrick done (although not updated the screenshots - too many)

File: app/client/components/updatableAmount.tsx [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    isApplyingUpdate: false,
+    currentAmount: this.initialAmount,
+    newAmount: this.initialAmount,
+    validationMessage: ""
+  };
+
+  public render(): React.ReactNode {
+    return (
+      <div css={wrappingContainerCSS}>
+        {this.state.isApplyingUpdate ? (
```
Comment (jacobwinch) [inline app/client/components/updatableAmount.tsx]: I find it pretty hard to follow code this type of code (where we use conditional logic and each expression is pretty big ReactNode). 

Perhaps extracting some nodes into consts and/or using some smaller functions would help to make this more readable?

File: app/server/server.ts [hunk 4]
```diff
@@ -44,7 +44,11 @@ const globals: Globals = {
 
 server.use(
   (req: express.Request, res: express.Response, next: express.NextFunction) => {
-    res.header("Cache-Control", "private");
+    // this header is VERY IMPORTANT and prevents caching (on both CDN and in browsers)
```
Comment (jacobwinch) [inline app/server/server.ts]: I would address this in its own PR - it doesn't seem related to the Contributions tab to me?

File: app/client/components/updatableAmount.tsx [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+import AsyncLoader from "./asyncLoader";
+import { Button } from "./buttons";
+import { wrappingContainerCSS } from "./productPage";
+
+enum MinOrMax {
+  minimum,
+  maximum
+}
+
+// min & max values based on https://github.com/guardian/support-frontend/blob/5aa237d/assets/helpers/contributions.js
```
[the remark this answers is not in the document]
Comment (jacobwinch) [reply app/client/components/updatableAmount.tsx]: Also, I quite like the support-frontend convention of separate this logic into helpers files - what do you think about doing something similar in this project?

File: app/client/components/updatableAmount.tsx [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+export class UpdatableAmount extends React.Component<
+  UpdatableAmountProps,
+  UpdatableAmountState
+> {
+  public initialAmount = this.props.subscription.nextPaymentPrice / 100.0;
+  public state = {
+    inEditMode: false,
+    isApplyingUpdate: false,
+    currentAmount: this.initialAmount,
+    newAmount: this.initialAmount,
```
Comment (jacobwinch) [inline app/client/components/updatableAmount.tsx]: It might be clearer if this was null to begin with?

File: app/client/components/updatableAmount.tsx [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+                currentAmount: this.state.newAmount
+              });
+              return null;
+            }}
+            loadingMessage={"Updating..."}
+            errorRender={() => {
+              this.setState({
+                isApplyingUpdate: false,
+                validationMessage:
+                  "Updating failed this time. Please try again later..."
```
Comment (jacobwinch) [inline app/client/components/updatableAmount.tsx]: We should add a card to analyse why this update might fail and whether this message is always going to be appropriate (definitely one for a follow up PR though)

File: app/server/server.ts (hunk 4 above)
Comment (twrichards) [reply app/server/server.ts]: moved it to https://github.com/guardian/manage-frontend/pull/140 (which also nicely covers one the remediation Trello card)

File: app/client/components/updatableAmount.tsx [hunk 8]
```diff
@@ -0,0 +1,287 @@
+import React, { ChangeEvent } from "react";
```
[the remark this answers is not in the document]
Comment (twrichards) [reply app/client/components/updatableAmount.tsx]: not really since this component is used to render the amount both in and out of edit mode, so it is an 'updatable amount'

File: app/client/components/updatableAmount.tsx (hunk 6 above)
Comment (twrichards) [reply app/client/components/updatableAmount.tsx]: not sure i agree, these state values need to start with the seed value from the API, mainly so the validation can show they need to change the value before they can confirm, see...
image

File: app/client/components/updatableAmount.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (twrichards) [reply app/client/components/updatableAmount.tsx]: added tests.

not sure I agree regarding moving out into global config file as that undermines the encapsulation of this self contained component. If in future we need these thresholds elsewhere in the application then that will warrant centralising them.

File: app/client/styles/fonts.ts [hunk 9]
```diff
@@ -328,3 +330,9 @@ export const fonts = `
 export const sans = `"Guardian Text Sans Web", "Helvetica Neue", Helvetica, Arial, "Lucida Grande", sans-serif`;
 export const serif = `"GuardianTextEgyptian", Georgia, serif`;
 export const headline = `GH Guardian Headline, ${serif}`;
+
+export const validationWarningCSS = {
```
[the remark this answers is not in the document]
Comment (twrichards) [reply app/client/styles/fonts.ts]: since its used in a few places and its purely typeface related CSS its a good enough home for it for now
